components/patient-data/impl/src/main/java/org/phenotips/data/internal/controller/FeaturesController.java

```diff
@@ -281,96 +297,205 @@ private JSONArray joinArrays(JSONArray jsonOne, JSONArray jsonTwo)
-            if (!feature.isPresent()) {
-                featureType = PhenoTipsFeature.NEGATIVE_PHENOTYPE_PREFIX + featureType;
-            }
-            if (featuresMap.keySet().contains(featureType)) {
-                featuresMap.get(featureType).add(feature.getValue());
+    @Override
+    public void save(@Nonnull final Patient patient, @Nonnull final PatientWritePolicy policy)
+    {
+        try {
+            final XWikiDocument docX = patient.getXDocument();
+            final XWikiContext context = this.xcontextProvider.get();
+            final BaseObject dataHolder = docX.getXObject(Patient.CLASS_REFERENCE, true, context);
+            final PatientData<Feature> features = patient.getData(getName());
+            if (features == null) {
+                if (PatientWritePolicy.REPLACE.equals(policy)) {
+                    clearFeatureData(docX, dataHolder, context);
+                }
             } else {
-                List<String> newFeatureType = new LinkedList<>();
-                newFeatureType.add(feature.getValue());
-                featuresMap.put(featureType, newFeatureType);
+                if (!features.isIndexed()) {
+                    this.logger.error(ERROR_MESSAGE_DATA_IN_MEMORY_IN_WRONG_FORMAT);
+                    return;
+                }
+                saveFeatures(docX, dataHolder, patient, features, policy, context);
             }
+        } catch (final Exception ex) {
+            this.logger.error("Failed to save features data: {}", ex.getMessage(), ex);
         }
+    }
 
-        // to reset values in the document null them first
-        for (String type : PHENOTYPE_PROPERTIES) {
-            data.set(type, null, context);
-        }
+    /**
+     * Saves specified {@code features} for {@code patient} according to the provided {@code policy}.
+     *
+     * @param docX the {@link XWikiDocument} object for the {@code patient}
+     * @param dataHolder the {@link BaseObject} for writing features
+     * @param patient the {@link Patient} of interest
+     * @param features a {@link PatientData} object containing feature data to be saved
+     * @param policy the {@link PatientWritePolicy} according to which data will be saved
+     * @param context the {@link XWikiContext}
+     */
+    private void saveFeatures(
+        @Nonnull final XWikiDocument docX,
+        @Nonnull final BaseObject dataHolder,
+        @Nonnull final Patient patient,
+        @Nonnull final PatientData<Feature> features,
+        @Nonnull final PatientWritePolicy policy,
+        @Nonnull final XWikiContext context)
+    {
+        final Stream<Feature> featureStream = PatientWritePolicy.MERGE.equals(policy)
+            ? buildFeaturesStream(load(patient), features)
+            : StreamSupport.stream(features.spliterator(), false);
+
+        clearFeatureData(docX, dataHolder, context);
+
+        featureStream
+            .map(feature -> saveFeatureData(docX, feature, context))
```

Comment:
In this case, I mostly wanted to take advantage of ```groupingBy```, since it's nice and short.
I haven't thought of this before, but the saving of feature data could also be moved to the following ```Collectors.mapping``` call, where i could save feature and return feature value. So that's 3 potential ways to do this. 

Implemented last, but open to discussion re: the best way to do this.